[WiP] aoide - External music library

uklotzde opened the pull request:
***Only test this PR if you are familiar with how to reset the database schema version in the `settings` table for reapplying any schema migrations that will precede integration of this foundational feature!!***

- [ ] Based on #2656  (DB v39)

Just a convenient starting point to demonstrate the current progress. The branch will be rebased on master frequently and the following instructions will be updated as required.

## Installation

### Mixxx build

Enabled with the feature flag `-DAOIDE=ON` in the CMake build.

### aoide Executable (Rust)

Requires a working version of [aoide](https://gitlab.com/uklotzde/aoide-rs) in the _program_ or _settings_ directory. A statically linked executable (Linux, x86_64) that works out of the box is available on the [release page](https://gitlab.com/uklotzde/aoide-rs/-/releases). Otherwise you need to install a Rust development environment and build it with `cargo build` or `cargo build --release` yourself.

I recommend placing the executable file in your personal Mixxx settings folder for testing purposes.

## Usage

### Startup/Shutdown

The backend service is started implicitly as a child process by Mixxx on startup and terminated on exit. Upon the first start it will create a new database *aoide.sqlite* in the settings directory, right next to *mixxxdb.sqlite*.

### Synchronization

Every *save* and *purge* action is propagated from Mixxx to aoide. You won't notice, because this is performed asynchronously in a separate thread. If you play tracks in Mixxx they will appear automatically in aoide after a while, because modifying the play count results in a *save* action eventually.

To initially populate the *aoide* database select some tracks in Mixxx and invoke *Send to aoide* from the metadata context menu. Repeat to enforce a synchronization. Unfortunately there is no progress dialog, so start experimenting with a small number of tracks.

### View

The feature in the side pane is empty after startup. You need to load a tree with prepared queries from a file (right click on feature title). An example file is provided in res/preferences/aoide_prepared_queries_example.json. 

Copy this file into your settings folder, because this is the default location when opening the file dialog. The syntax and schema of this file is not fixed yet, so don't spend too much work into tuning it. Don't forget to replace your copied version with an updated version from the repo, if necessary.

![aoide](https://user-images.githubusercontent.com/1474154/64921499-3ee0ea00-d7c4-11e9-88fb-0e4960dce41d.png)

***This is just a dumb, read-only view and all track interactions are forwarded to Mixxx.***

### Monitoring & Logging

The Mixxx log level is applied to aoide and all aoide logs are picked up by Mixxx. By setting the log level to _debug_ aoide will generate many logs in your Mixxx log file. This will reveal the progress of batch operations and all the inter-process communication.

## Tagging

One of the distinctive features of aoide is the extensible tagging system that allows to support multi-valued fields. Artists, titles, and some other fields are stored as predefined properties, but anything else like genres, comments, grouping, lyrics, ... must be mapped to custom tags.

### Genre tags

Mixxx only supports a single genre per track. But you could encode multiple genres into this field by using a separator like " - " (*space* *dash* *space*): "Pop - Rock".

In aoide the genre is stored in a dedicated tag with the *facet* "genre". This is just a convention. Multiple tags with different *labels* like "Pop" and "Rock" can be assigned to a track. The order of the genres is determined by their *score* between 0.0 and 1.0. If missing the implicit default score for any tag is 1.0.

The export from Mixxx to aoide splits the *genre* field into separate tags by using a customizable separator and an attenuation value for assigning a diminishing score to sub-genres. The default values are not stored in the Mixxx configuration file, but they would look like in this example:

```
[aoide]
multiGenreSeparator " - "
multiGenreAttenuation 0.75
```

The separator is enclosed in single ('...') or double ("...") quotes to avoid that leading and trailing whitespace gets cut off when editing the .cfg file. The 1st (= main) genre gets a score of 0.75^0 = 1, the 2nd genre gets a score of 0.75^1 = 0.75, the 3rd genre a score of 0.75^2 = 0.5625 and so on.

In our simple example the following tags would be exported to aoide:
- facet = "genre", label = "Pop", score = 1.0
- facet = "genre", label = "Rock", score = 0.75

When receiving tracks from aoide both the separator and the assigned score is used to restore the single-valued genre field for the view. The genre tags are sorted in descending order by their score and then their labels are joined with the separator from the configuration.

The `MultiGenreTagger` implements this bidirectional mapping.

## Side effects

**None** ...as far as I'm aware! You can safely switch back to master after testing any time you want. The Mixxx database is not affected in any way.

You may need or want to delete the aoide SQLite database from time to time. I will not provide any schema migrations until v1.0 has been released. Use your Mixxx library as the source for repopulating the aoide library from scratch.

## Path forward

Some extensions need to be merged first. After all preliminary PRs have been merged this PR should finally only contain commits that add the actual integration.

uklotzde committed:
Add EncodedUrl

daschuer commented:
Thank you very much for all the love and time you put into this. It looks very promising. :-)

Skimming through the code the use of QPointer jumps up. 

Qt docs: 
> QPointers are useful whenever you need to store a pointer to a QObject that is owned by someone else, and therefore might be destroyed while you still hold a reference to it. You can safely test the pointer for validity.  

Agent() for example holds Subsystem as QPointer<Subsystem> which is correct because it is a Pointer owned by someone else. Reading only the Agent() code, the it tells me QPointer can become nullptr at any time, so I would expect to use null checks for every access. 

In this case it we can omit the checks because the calling code guarantees that Subsystem lives longer. 
QPointer implies the opposite. Unfortunately we can also not use a plain pointer, because with this, the function grantees to not store the pointer which is also not true. 

So I think we have a gap here between ```QPointer<T> parented_ptr::toWeakRef()``` and ```T* parented_ptr::get()``` we should close. 

https://en.cppreference.com/w/cpp/experimental/observer_ptr could be an alternative for our purpose but the lifetime is also not documented. 

gsl::not_null<T*> can also be used ...
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-nullptr

But how about create an own pointer type for documentation purpose: 
outliving_ptr

uklotzde commented:
@daschuer I've deliberately used QPointer instead of plain or other managed pointers to quickly detect any lifetime conflicts between Qt-managed objects. A crash by dereferencing a nullptr is the most reliable way to reveal those dependencies. QPointer guarantees that we don't end up with dangling pointers.

This behavior can be tuned and adjusted after the design of all subsystem components and their relationships have been finalized.

daschuer commented:
Yes, right, during debugging a crash from object de-referenced from a nullptr is more easy to find than a crash from a dangling pointer. Unfortunately it is still a crash and a QPointer does not come for free, it installs the reference counting calling potentially locking new().  

These crashes are a result from a misuse of pointers, probably caused by misinterpreting existing code.
The idea was to prevent this before the crash happens by unambiguous define the live time of the referenced Object. For me QPointer does the opposite. 

If we can effort the reference counting overhead, we may just inherit a outliving_ptr from QPointer for that purpose. But that should not be necessary because In many cases it can be easily guaranteed that the outliving_ptr guarantee is valid. If this is not obvious, QPointer with manual null check can be used. or a parent-less QObject managed by a QSharedPointer.

Other Ideas?

uklotzde commented:
I have reduced the usage of QPointer to unowned members. It works exactly as documented and intended, i.e. an implicit weak pointer.

uklotzde commented on src/library/aoide/trackcollection.cpp:
```diff
@@ -0,0 +1,76 @@
+          m_trackCollection(trackCollection),
+          m_trackLoader(new AsyncTrackLoader(trackCollection, this)),
+          m_subsystem(new Subsystem(userSettings, m_trackLoader, this)),
+          m_agent(new Agent(m_subsystem, this)) {
```
No. The agent coordinates the selection of an active collection without user interaction. Signals are needed because the whole process is asynchronous.

daschuer replied:
I don't get that from the code.
It looks like as if it is possible to inline all agent code. Is there a worker thread involved? Where is it?

uklotzde replied:
I've added a comment to `Agent`. All child components of `TrackCollection` (`Agent`, `Subsystem, `AsyncTrackLoader`) are running on the UI event loop as expected.

The worker thread and all its components are controlled internally by `Subsystem` that acts as a facade. This class also handles startup and shutdown of the child process. The name has been chosen deliberately.

uklotzde commented on src/library/aoide/domain/cue.cpp:
```diff
@@ -0,0 +1,148 @@
+    switch (cue.getType()) {
+    case Cue::CUE:
+    case Cue::LOAD:
+        setStart(cue.getPosition() * cuePositionToMillis);
```
Not possible, because unfortunately cue position and length are measured in sample offsets, assuming that channel samples are interleaved. The latter is one of the main design flaws that complicates the engine code in so many places. Moreover, these cue properties depend on the context, namely the sample rate and channel count of the PCM signal.

```
cuePositionToMillis = 1000.0 / (sampleRate * channelCount)
```

aoide fixes this by storing the values in plain time units, independent from the properties of the underlying PCM signal.

daschuer replied:
Ha, my bad. Can you explain this inline.

uklotzde replied:
I have added a comment in the `TrackExporter` where this value is calculated. Cue points of tracks with an invalid sample rate or number of channels cannot be exported. It only happens for corrupt or unsupported files.

daschuer commented on src/library/aoide/domain/cue.h:
```diff
@@ -0,0 +1,35 @@
+class AoidePositionMarker : public AoideJsonObject {
+  public:
+    AoidePositionMarker(const Cue& cue, double cuePositionToMillis);
+    explicit AoidePositionMarker(QJsonObject jsonObject = QJsonObject())
```
I think it would be better to name the file like the class

uklotzde commented:
I've improved the timeout and error handling so that network request processing doesn't get stuck. This is essential, because write requests are queued and processed sequentially to guarantee that all modifications are applied in order.

uklotzde commented:
Mixxx now saves the file path with the last tree view configuration that has been loaded and automatically restores it on startup. No need to store this information in the database.

uklotzde commented:
v0.5.0 now actively prevents adding different tracks with the same URIs into a single collection. This should help to detect bugs in the Mixxx external collection synchronization code like those fixed by #2404.

Be-ing commented:
> You need to load a tree with prepared queries from a file (right click on feature title). An example file is provided in res/aoide/example_prepared_queries.json.

Please update the first post. I got confused because you have moved this JSON file to res/preferences.

uklotzde commented:
> > You need to load a tree with prepared queries from a file (right click on feature title). An example file is provided in res/aoide/example_prepared_queries.json.
> 
> Please update the first post. I got confused because you have moved this JSON file to res/preferences.

Yes, sorry, forgot about it. After the custom tags are also loaded from a file i have move both into a common place.

uklotzde commented:
@Be-ing I have also deleted the "Custom tags" section. This does no longer apply after I removed the hand-written pseudo-parse. This will be replaced by proper custom tags support in Mixxx soon that will be compatible with a aoide tagging scheme.

Be-ing commented:
> The feature in the side pane is empty after startup. You need to load a tree with prepared queries from a file (right click on feature title).

This was quite confusing to me at first. I expected clicking "aoide" in the tree to show all tracks in Aoide, just like clicking on Tracks.

uklotzde commented:
> > The feature in the side pane is empty after startup. You need to load a tree with prepared queries from a file (right click on feature title).
> 
> This was quite confusing to me at first. I expected clicking "aoide" in the tree to show all tracks in Aoide, just like clicking on Tracks.

Showing all tracks without any filtering and sorting doesn't make sense. It would display the first arbitrary 250 tracks. We have to drop this "load the whole database into memory" concept, that doesn't fit here.

Sorting should be configurable independent of the view. Example: In Mixxx I perfer to have the main view ordered by album and then by date added in descending order, i.e. to show the new tracks. This always requires 3 clicks (interrupted by long UI pauses to finish the "load the whole database task" although I discard 2 of them immediately) if I need to restore this view, very cumbersome. In the aoide view this is just  a prepared query that I can save and recall whenever I need it.

Be-ing commented:
> Showing all tracks without any filtering and sorting doesn't make sense.

Yes, showing everything with no filtering nor sorting makes no sense. But we need a default query with no filtering and only sorting for the user to scroll through everything. I don't care much what the default sort order is.

uklotzde commented:
> > Showing all tracks without any filtering and sorting doesn't make sense.
> 
> Yes, showing everything with no filtering nor sorting makes no sense. But we need a default query with no filtering and only sorting for the user to scroll through everything. I don't care much what the default sort order is.

The "Recently Added" query is ideal for this purpose. In conjunction with "Recently Modified" this could be provided as a built-in default that does not need to be loaded on a fresh install. We could either bundle or generate a default configuration for this purpose.

Is suggest that the main view is reserved for an overview of the feature, not for displaying tracks. It could include quick links and diagnostic/statistical information.

Be-ing commented:
This branch currently does not build:
```
/home/be/sw/mixxx/src/track/trackrecord.cpp:108:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refConductor(), importedTrackInfo.getConductor());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:109:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refDiscNumber(), importedTrackInfo.getDiscNumber());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:110:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refDiscTotal(), importedTrackInfo.getDiscTotal());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:111:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refEncoder(), importedTrackInfo.getEncoder());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:112:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refEncoderSettings(), importedTrackInfo.getEncoderSettings());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:113:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refISRC(), importedTrackInfo.getISRC());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:114:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refLanguage(), importedTrackInfo.getLanguage());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:115:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refLyricist(), importedTrackInfo.getLyricist());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:116:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMood(), importedTrackInfo.getMood());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:117:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMovement(), importedTrackInfo.getMovement());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:118:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMusicBrainzArtistId(), importedTrackInfo.getMusicBrainzArtistId());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:119:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMusicBrainzRecordingId(), importedTrackInfo.getMusicBrainzRecordingId());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:120:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMusicBrainzReleaseId(), importedTrackInfo.getMusicBrainzReleaseId());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:121:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refMusicBrainzWorkId(), importedTrackInfo.getMusicBrainzWorkId());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:122:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refRemixer(), importedTrackInfo.getRemixer());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:123:20: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotEmpty(mergedTrackInfo.refSeratoTags(), importedTrackInfo.getSeratoTags());
                   ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:124:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refSubtitle(), importedTrackInfo.getSubtitle());
                  ^
/home/be/sw/mixxx/src/track/trackrecord.cpp:125:19: error: use of undeclared identifier 'mergedTrackInfo'
    copyIfNotNull(mergedTrackInfo.refWork(), importedTrackInfo.getWork());
                  ^
18 errors generated.
gmake[2]: *** [CMakeFiles/mixxx-lib.dir/build.make:4955: CMakeFiles/mixxx-lib.dir/src/track/trackrecord.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:1562: CMakeFiles/mixxx-lib.dir/all] Error 2
gmake: *** [Makefile:180: all] Error 2
```

uklotzde commented:
Daniel's changes seem to have broken the __EXTRA_METADATA__ feature. I was assuming this flag is enabled for our CI builds, doesn't seem so.

Be-ing commented:
This segfaults on startup now. I think there is a recursive function call because this line appears thousands of times in the backtrace to the point that it completely fills my console's buffer:
```
Debug [Main]: WTrackTableView::loadTrackModel() AnalysisLibraryTableModel(0x9d07240)
Process 44051 stopped
* thread #24, name = 'BrowseThread', stop reason = signal SIGSEGV: address access protected (fault address: 0x7fff11afbff8)
    frame #0: 0x0000000000bfe7c8 mixxx`mixxx::taglib::riff::importTrackMetadataFromTag(pTrackMetadata=0x00007fff122fa090, tag=0x00007ffef801b410) at trackmetadata_riff.cpp:13:5
   10  	    TrackMetadata* pTrackMetadata,
   11  	     const TagLib::RIFF::Info::Tag& tag) {
   12  	    // Just delegate to the common import function
-> 13  	    importTrackMetadataFromTag(pTrackMetadata, tag);
   14  	}
   15  	
   16  	} // namespace riff

...

frame #256971: 0x0000000000bfe7cd mixxx`mixxx::taglib::riff::importTrackMetadataFromTag(pTrackMetadata=0x00007fff122fa090, tag=0x00007ffef801b410) at trackmetadata_riff.cpp:13:5
```

uklotzde committed:
Add a method for writing custom messages into the log

afalout commented:
Hello,

First, thank you for sharing your work!

Was trying to test latest PR on Windows;
Complied aoide.exe 0.7.4
Downloaded last Mixxx build from https://github.com/mixxxdj/mixxx/suites/2193964997/artifacts/45264670
(noting that aoide.exe was already in %LOCALAPPDATA%\Mixxx\ when I tried to copy it there - I'm assuming "cargo build" put it there?)
tried running aoide from cmd, all seems OK
When trying to run Mixxx, window opens, but after a few seconds exits:
C:\Program Files\Mixxx>mixxx 
C:\Program Files\Mixxx>Loading resources from  "C:/Program Files/Mixxx/" 
Configuration file is at version "2.3.0-beta" instead of the current 2.4.0-alpha-pre 
BroadcastSettings - Found 1 profile(s) 
Loading resources from  "C:/Program Files/Mixxx/" 
Found and will use default keyboard mapping "C:/Program Files/Mixxx/keyboard/en_US.kbd.cfg" 
Loading resources from  "C:/Program Files/Mixxx/" 
DEBUG ASSERT: "!formatInterface().isEmpty()" in function class QString __cdecl mixxx::hid::DeviceInfo::formatUsage(void) const at ..\src\controllers\hid\hiddevice.cpp:76 
C:\Program Files\Mixxx> 

Did I missed any prerequisite steps?
Thanks!

uklotzde commented:
> Did I missed any prerequisite steps?
> Thanks!

There is currently heavy work in progress with more breaking changes on the aoide side. This branch only works with the current development version.

Please note that this branch add 2 database migrations and you should be aware of the consequences if you try it on an existing Mixxx database. If in doubt better use a throwaway database, e.g. use a copy of your main database.

Be-ing commented:
>  DEBUG ASSERT: "!formatInterface().isEmpty()" in function class QString __cdecl mixxx::hid::DeviceInfo::formatUsage(void) const at ..\src\controllers\hid\hiddevice.cpp:76

It looks like you ran into an unrelated bug in Mixxx.

uklotzde commented:
> > DEBUG ASSERT: "!formatInterface().isEmpty()" in function class QString __cdecl mixxx::hid::DeviceInfo::formatUsage(void) const at ..\src\controllers\hid\hiddevice.cpp:76
> 
> It looks like you ran into an unrelated bug in Mixxx.

Oh, right, this is unrelated.

afalout commented:
> > > DEBUG ASSERT: "!formatInterface().isEmpty()" in function class QString __cdecl mixxx::hid::DeviceInfo::formatUsage(void) const at ..\src\controllers\hid\hiddevice.cpp:76
> > 
> > 
> > It looks like you ran into an unrelated bug in Mixxx.
> 
> Oh, right, this is unrelated.

Thanks for your  reply; So is there a working current aoide build I could test?
Thanks

Be-ing commented:
I have a few preliminary comments from a quick glance at the code.

What is the difference between src/aoide/json and src/aoide/web? These seem to be thin wrapper classes around the Aoide HTTP API so I suppose there isn't much to document about them apart from referring to the Aoide documentation. Come to think of it, it would be helpful to add a link to the Aoide documentation for each wrapper class in Mixxx.

Please add documentation to each new class with a brief description of its responsibilities and how it interacts with other parts of Mixxx. See #2618 for example.

Maybe rename aoide::Subsystem to aoide::Manager for consistency with the rest of Mixxx?

Where do you suggest to start reviewing in 20,000 lines of new code?

uklotzde commented:
@Be-ing This PR is not ready for review yet. It includes changes for at least 3 preliminary PRs: File info rework + custom tags (#2656) + metadata synchronization timestamps.

Be-ing commented:
Yes, I was only looking at the last commit.

uklotzde commented:
This last commit has to be split into individual features that we can merge step by step:

- Foundation
- JSON wrappers/shims
- Network tasks (one by one)

But I will not start splitting it before the preliminary work is done. I already again invested hours to prepare splitting of the basic file info stuff that was mixed up while resolving some merge conflicts sometimes.

Be-ing commented:
> But I will not start splitting it before the preliminary work is done. I already again invested hours to prepare splitting of the basic file info stuff that was mixed up while resolving some merge conflicts sometimes.

Yes, that's reasonable. There's no need to wait for that to document the code though.

uklotzde committed:
Save custom tags in file tags

afalout commented:
@uklotzde - may I please ask, is this ready to be tested?
Thanks!

uklotzde commented:
> @uklotzde - may I please ask, is this ready to be tested?
> Thanks!

Permanently enabled in my custom builds, no guarantees though. You should be aware about the consequences and how to deal with them, i.e. database schema changes. Only works with the current development version.

afalout commented:
@uklotzde - thanks. 
1) I pulled fresh git for aoide and compiled, copied exe to %LOCALAPPDATA%\Mixxx\ 
2) I removed previous Mixxx, downloaded installer from your last pull request, installed.
3) When I run Mixxx, aoide.exe is not started; aoide.sqlite is not created
4) There is no Aoide menu item on the left of the browser (as per your screenshot in instructions); however, there is a "Custom Tags" item in the context menu when I right-click on the track.
5) This allows me to "tick"/"check" existing "facets", and create new ones. Nice.
However....
a) Where is the data going? Where is the Aoide database? 
b) How is this working, considering that aoide.exe is not running? I also dont see any mention of "aoide" in logs...
c) I see no way to create "facets" that would allow me to store descriptive or numeric data - for example, DiscogsReleaseID=123456ABC
Thanks!

Be-ing commented:
AFAIK you're the first to test this on Windows. I suspect the code is failing because it does not account for the `.exe` suffix of the binary on Windows: https://github.com/mixxxdj/mixxx/pull/2282/files#diff-a31a5f22d4721d60a6b8c33fa2ff2bfbaf2d9dcf619fb200e2f1ea28f6927ce8R50

afalout commented:
Thanks @Be-ing  - so there is nothing I can do?
But should this at lease write something to the logs like "aoide executable not found" or similar?
Would it work if I start aoide.exe manually?
Cheers

uklotzde commented:
> Thanks @Be-ing - so there is nothing I can do?
> But should this at lease write something to the logs like "aoide executable not found" or similar?
> Would it work if I start aoide.exe manually?
> Cheers

Sorry, I didn't have time to keep this PR up-to-date recently.

You should have seen an info log in Mixxx if the executable was not been found:

```
    QFileInfo cmdFile(command);
    if (!cmdFile.exists()) {
        kLogger.info()
                << "Executable file not found"
                << command;
        command = kExecutableName;
    }
```

I have adjusted the executable file name depending on the platform. No guarantees, because I currently don't have a Windows development environment at hand.

afalout commented:
@uklotzde :

> You should have seen an info log in Mixxx if the executable was not been found:
> 
> ```
>     QFileInfo cmdFile(command);
>     if (!cmdFile.exists()) {
>         kLogger.info()
>                 << "Executable file not found"
>                 << command;
>         command = kExecutableName;
>     }

I can confirm most definitely there is not work "Executable" in log file I attached above. Perhaps on Windows this code is not even invoked?

> 
> I have adjusted the executable file name depending on the platform. No guarantees, because I currently don't have a Windows development environment at hand.

Tied installing last Mixxx build from this PR on Ubuntu 20.10, but it depends on libprotobuf-lite17, and only libprotobuf-lite23 is available on 20.10

Thanks for your help.

afalout commented:
Also - compiling current aoide git on Ubuntu fails with:

   Compiling serde_urlencoded v0.7.0
   Compiling diesel v1.4.6
   Compiling aoide-core v0.0.0 (/home/andrej/src/aoide/aoide-rs/core)
error[E0658]: use of unstable library feature 'clamp'
   --> core/src/music/time/mod.rs:290:53
    |
290 |         let measure_number = measure_offset.floor().clamp(
    |                                                     ^^^^^
    |
    = note: see issue #44095 <https://github.com/rust-lang/rust/issues/44095> for more information

error[E0658]: use of unstable library feature 'clamp'
  --> core/src/tag/mod.rs:53:22
   |
53 |         value.into().clamp(Self::min_value(), Self::max_value())
   |                      ^^^^^
   |
   = note: see issue #44095 <https://github.com/rust-lang/rust/issues/44095> for more information

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
error: could not compile `aoide-core`.

Swiftb0y commented:
`clamp` got stabelized in rustc 1.50.0, update your rust installation (assuming you are using rustup run `rustup update`), to at least 1.50.0 (1.51.0 is the newest version at the time of writing).

uklotzde commented:
> `clamp` got stabelized in rustc 1.50.0, update your rust installation (assuming you are using rustup run `rustup update`) to at least 1.50.0 (1.51.0 is the newest version at the time of writing).

Confirmed. I always update to the latest Rust version and start using the new features.

afalout commented:
Thanks - I upgraded Rust and installed libprotobuf-lite17 from Focal repo.

Still, no Aoide in browser tree, aoide executable was not started from /home/<user>/.mixxx/, and there is no word "Executable" in debug level log:

mixxx --log-level debug > mixxx-debug.log 2>&1
grep -i executable mixxx-debug.log 
grep -i aoide mixxx-debug.log 
grep -i executable /home/andrej/.mixxx/*.log*
grep -i aoide /home/andrej/.mixxx/*.log*

What am I doing wrong?
Thanks

afalout commented:
I must say I am completely confused with what I am seeing; I am looking the behavior that is described on top of this page (since my repeated attempts did not prompt anyone to say that I should not),  but what I see when I do manage to open application looks to me as what I guess  this PR is supposed to do:

![Screenshot_20210425_200017](https://user-images.githubusercontent.com/4283676/115985790-9d598580-a601-11eb-94bb-fb8b2eeef305.png)

So can someone please confirm that the instructions are still valid? 
Is there supposed to be an Aoide entry in browser menu? 
Is aoide executable still needed? 
Is it supposed to be started by Mixxx and running? 
Is there supposed to be an aoide.sqlite database created somewhere? 
How is the functionality under "Custom tags" in my screenshot working (since none of the above is happening), and where is data stored? 
Or is functionality I see performed by code completely unrelated to this PR?

Thanks

uklotzde committed:
Add aoide integration

uklotzde committed:
aoide: Mount as external track collection in read-only mode

uklotzde committed:
Add faceted custom tags foundation with JSON serialization

uklotzde committed:
UI: Reword options for track metadata sync with file tags

uklotzde committed:
Extract utility functions for parsing JSON data

uklotzde committed:
CustomTags: Log and report parsing errors

uklotzde committed:
DB: Update track metadata from file tags if outdated

When loading a track from the database check if the underlying file
has been modified. If the time stamp is newer than the timestamp of
the last synchronization then update track metadata from file tags.

But only if export of track metadata into file tags has been enabled
by the user to reduce surprising side-effects to a minimum. Introducing
a new configuration option for this feature has deliberately been avoided
to reduce the number of possible configurations. Either you want to
keep your file tags synchronized or not, no exceptions!

# Conflicts:
#	src/sources/metadatasourcetaglib.cpp

uklotzde committed:
Custom tags: Refine and extend documentation

uklotzde committed:
Custom tags: Rename member function

uklotzde committed:
Tag facet/label/score: clampValue() -> convertIntoValidValue()

uklotzde committed:
Tag facet/label: Trade constexpr for validation

Use debug assertions to validate the string values, even though that
precludes constexpr.

uklotzde committed:
Custom tags: Split source files

uklotzde committed:
TagFacet: Enable definition of static, non-validated constants

uklotzde committed:
Add tool tips to sidebar tree items

uklotzde committed:
Explain the purpose of facet identifiers

uklotzde committed:
Fix clazy warnings

uklotzde committed:
Restrict the alphabet of tag facets

uklotzde committed:
Update tag facet doc comment

uklotzde committed:
Renaming: TagFacet -> TagFacetId

uklotzde committed:
Tags: Add constants for predefined facet identifiers

uklotzde committed:
Fix naming of constant

uklotzde committed:
Add more predefined tag facets and labels

uklotzde committed:
Rename CustomTags as Facets

uklotzde committed:
Populate custom tags menu from selected/all tracks

uklotzde committed:
Undo: Save custom tags in file tags

uklotzde committed:
Library preferences: Reword file synchronization options

uklotzde committed:
Track source sync: Log possible reasons for runtime errors

uklotzde committed:
Remove redundant "Tag" prefix from type names

uklotzde committed:
Only re-import Serato tags if Serato metadata export is enabled

uklotzde committed:
Library Prefs: Reword and explain Serato metadata synchronization

uklotzde committed:
Add missing const qualifiers in test classes

uklotzde committed:
Parse query terms and search for tag facets and labels

Only tag labels are filtered, facets and scores are ignored. The
resulting subselect (if not empty) is added to the main query with
OR if the query contains no tag-specific search terms. If the query
contains at least one token "label:<text>" then AND will be used.

uklotzde committed:
Add utility functions and accompanying types for tag labels

uklotzde committed:
Restrict facet identifiers to start with a lowercase alphabetic letter

uklotzde committed:
Add custom tags

uklotzde committed:
aoide: Workaround for spurious SIGSEGV crashes

uklotzde commented:
Closing due to lack of motivation. Please do not reopen!

afalout commented:
Such a shame to give up after a great effort. I had high hopes for this very much needed functionality. Is there anything that can be done to help you continue this work? 
Thanks, Andrej

uklotzde commented:
I will maintain a stripped down Mixxx integration layer for personal use. But I won't sink more time into the Mixxx integration.

ronso0 commented:
:( what a pity
Admittedly I didn't bother testing this because I'm served well with my abuse of comments, and don't use crates anymore.
Nevertheless, I appreciated your continuous work, hoping this could be the foundation of a new library design (especially the search part).
